File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    if (!prices.length) return 0;
    let minPrice = prices[0], profit = 0;
    for (let i = 1; i < prices.length; i++) {
        profit = Math.max(profit, prices[i] - minPrice);
        minPrice = Math.min(minPrice, prices[i]);
    }
    return profit;
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int maxSum = 0; int minVal = prices[0];

        for(int i = 1; i < prices.length; i++) {
            maxSum = Math.max(maxSum,
                prices[i] - minVal 
            );
            minVal = Math.min(minVal, prices[i]);
        }
        return maxSum;
    }
}